src/SimpleJson/SimpleJson.cs

```diff
@@ -1412,6 +1415,21 @@ public virtual object DeserializeObject(object value, Type type)
                             dict.Add(kvp.Key, DeserializeObject(kvp.Value, valueType));
 
                         obj = dict;
+
+#if SIMPLE_JSON_READONLY_COLLECTIONS
+                        // Wrap dictionary in a ReadOnlyDictionary<,>
+                        var genericTypeDefinition = type.GetGenericTypeDefinition();
+                        if (genericTypeDefinition == typeof(IReadOnlyDictionary<,>) ||
+                            genericTypeDefinition == typeof(ReadOnlyDictionary<,>))
+                        {
+                            var ctorType = typeof(IDictionary<,>).MakeGenericType(keyType, valueType);
+                            var genericReadonlyType = typeof(ReadOnlyDictionary<,>).MakeGenericType(keyType, valueType);
+                            var ctor = ReflectionUtils.GetContructor(genericReadonlyType, new Type[] { ctorType });

```

Comment:
Shouldn't directly use `ReflectionUtils.GetConstructor` as it doesn't cache. `ConstructorCache` should be used. Using cache increases the 2nd and nth performance drastically.

`GetConstructor` can internally call `GetConstructorByReflection` or `GetConstructorByExpression` depending on the settings. The performance impact will be seen on `GetConstructorByExpression` since it needs to compile the lambda expression. Compiling lamda expression are slow and hence is only done once when using the cache. https://github.com/facebook-csharp-sdk/simple-json/blob/f739f0b2924bcfa33204e86b8b73aa98be68abac/src/SimpleJson/SimpleJson.cs#L1861
Here is the (cryptic) benchmark that shows a huge performance gain. https://gist.github.com/prabirshrestha/2433886#file-benchmark-on-net-4-client-profile-L17

ctor compiled lambda
00:00:00.0051271 // first run
00:00:00.0000074 // second run

benchmark code can be found at https://github.com/facebook-csharp-sdk/reflection-utils/blob/e1ef1b8dca63efb0826e85bc4987f1685537a7ca/src/ReflectionUtils/ReflectionUtilsBenchmarks.cs